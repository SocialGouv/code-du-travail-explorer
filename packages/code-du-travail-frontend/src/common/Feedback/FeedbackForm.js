import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { Alert } from "@cdt/ui-old";
import { Button, theme } from "@socialgouv/react-ui";

import { feedbackUrl } from "./feedback.service";

const motifLabels = [
  "Les informations proposées ne sont pas à jour",
  "Les informations me paraissent erronées",
  "Les informations me paraissent incomplètes",
  "Les informations me paraissent hors-sujet",
  "Je n'ai pas trouvé la réponse à ma question"
];

class FeedbackForm extends React.Component {
  static propTypes = {
    query: PropTypes.string,
    title: PropTypes.string,
    isSatisfied: PropTypes.bool.isRequired,
    onSubmit: PropTypes.func.isRequired,
    sourceFilter: PropTypes.string.isRequired,
    sourceType: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
  };
  static defaultProps = {
    query: "",
    title: ""
  };
  state = {
    status: "", // "" | "sending" | "error"
    motif: null,
    message: "",
    question: this.props.query,
    email: ""
  };
  timeoutId = null;
  texteareaRef = React.createRef();

  static getDerivedStateFromProps(props) {
    return {
      motif: !props.isSatisfied ? motifLabels[0] : null
    };
  }

  componentDidMount() {
    this.texteareaRef.current.focus();
  }

  componentWillUnmount() {
    clearTimeout(this.timeoutId);
  }

  inputChange = event => {
    this.setState({
      [event.target.name]: event.target.value
    });
  };

  onSubmit = async event => {
    event.preventDefault();
    const { question, motif, message } = this.state;
    if (message.replace(/\s+/, "").trim() === "") {
      alert("Merci de compléter le formulaire");
      return;
    }
    this.setState({ status: "sending" });

    const data = {
      sourceFilter: this.props.sourceFilter,
      isSatisfied: this.props.isSatisfied,
      message,
      motif,
      query: this.props.query,
      sourceType: this.props.sourceType,
      subject: question,
      title: this.props.title,
      url: document.location.href,
      userAgent: typeof navigator !== "undefined" && navigator.userAgent
    };
    try {
      await this.props.onSubmit(data);
      this.setState({
        status: "",
        message: "",
        motif: motifLabels[0]
      });
    } catch (error) {
      this.setState({ status: "error" });
      this.timeoutId = setTimeout(() => this.setState({ status: "" }), 3000);
    }
  };

  getAlert() {
    switch (this.state.status) {
      case "sent":
        return <Alert variant="success">Message bien envoyé !</Alert>;

      case "error":
        return (
          <Alert variant="warning">Impossible d’envoyer votre message</Alert>
        );

      default:
        return null;
    }
  }

  render() {
    const { query, url, sourceType, isSatisfied } = this.props;

    return (
      <StyledForm action={feedbackUrl} onSubmit={this.onSubmit}>
        <p>
          <strong>Merci pour votre réponse !</strong>
        </p>
        <p>
          Vous souhaitez nous aider à améliorer le Code du travail numérique ?
          Laissez-nous un commentaire :
        </p>
        <input type="hidden" name="question" value={query} />
        <input type="hidden" name="url" value={url} />
        <input type="hidden" name="sourceType" value={sourceType} />
        {!isSatisfied && (
          <StyledInput
            as="select"
            name="motif"
            onChange={this.inputChange}
            onBlur={this.inputChange}
            value={this.state.motif}
          >
            {motifLabels.map((label, i) => (
              <option key={`motif-${i}`} value={label}>
                {label}
              </option>
            ))}
          </StyledInput>
        )}
        <StyledInput
          as="textarea"
          ref={this.texteareaRef}
          name="message"
          placeholder="Les informations ..."
          onChange={this.inputChange}
          value={this.state.message}
        />

        <FormAction>
          <Button variant="primary" disabled={this.state.status === "sending"}>
            Envoyer mon commentaire
          </Button>
          <Status>{this.getAlert()}</Status>
        </FormAction>
      </StyledForm>
    );
  }
}

export { FeedbackForm };

const { spacing } = theme;

const StyledForm = styled.form`
  display: flex;
  width: 100%;
  flex-direction: column;
  align-items: stretch;
`;

const StyledInput = styled.input`
  margin-bottom: ${spacing.medium};
`;

const FormAction = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

const Status = styled.div`
  flex-basis: 100%;
  margin: 0;
  margin-top: ${spacing.medium};
`;
