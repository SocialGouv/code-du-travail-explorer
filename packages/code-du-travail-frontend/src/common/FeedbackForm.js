import React from "react";
import PropTypes from "prop-types";
import Link from "next/link";
import { Alert } from "@cdt/ui";
import { getRouteBySource } from "../sources";
import { feedbackUrl } from "./feedback.service";

const motifLabels = [
  "Les informations proposées ne sont pas à jour",
  "Les informations me paraissent erronées",
  "Les informations me paraissent incomplètes",
  "Les informations me paraissent hors-sujet",
  "Je n'ai pas trouvé la réponse à ma question"
];

class FeedbackForm extends React.Component {
  static propTypes = {
    query: PropTypes.string,
    url: PropTypes.string,
    results: PropTypes.arrayOf(PropTypes.object),
    onSubmit: PropTypes.func.isRequired
  };
  static defaultProps = {
    query: "",
    url: "",
    results: []
  };
  state = {
    status: "", // "" | "sending" | "sent" | "error"
    motif: motifLabels[0],
    message: "",
    question: this.props.query || "",
    email: ""
  };
  timeoutId = null;
  texteareaRef = React.createRef();

  componentDidMount() {
    this.texteareaRef.current.focus();
  }

  componentWillUnmount() {
    clearTimeout(this.timeoutId);
  }

  inputChange = event => {
    this.setState({
      [event.target.name]: event.target.value
    });
  };

  onSubmit = event => {
    event.preventDefault();
    const { question, motif, message, email } = this.state;
    if (!email || email.indexOf("@") === -1) {
      alert("Merci de compléter le formulaire");
      return;
    }

    this.setState({ status: "sending" });

    const data = {
      motif,
      message,
      email,
      url: document.location.href,
      userAgent: typeof navigator !== "undefined" && navigator.userAgent,
      subject: question
    };
    this.props
      .onSubmit(data)
      .then(() => {
        this.setState({
          status: "sent",
          email: "",
          message: "",
          motif: motifLabels[0]
        });
      })
      .catch(() => {
        this.setState({ status: "error" });
      })
      .then(() => {
        this.timeoutId = setTimeout(() => this.setState({ status: "" }), 3000);
      });
  };

  getAlert() {
    switch (this.state.status) {
      case "sent":
        return <Alert success>Message bien envoyé !</Alert>;

      case "error":
        return <Alert warning>Impossible d&apos;envoyer votre message</Alert>;

      default:
        return null;
    }
  }
  render() {
    const { results, query, url } = this.props;

    return (
      <form
        action={feedbackUrl}
        className="feedback__form"
        onSubmit={this.onSubmit}
      >
        <input
          type="text"
          name="question"
          value={this.props.query}
          onChange={this.inputChange}
          className="feedback__input"
          disabled
        />
        <input
          type="hidden"
          name="url"
          value={document ? document.location.href : url}
        />
        {results &&
          results.length > 0 && (
            <React.Fragment>
              <h2 className="section__subtitle" style={{ alignSelf: "center" }}>
                Les réponses qui pourraient vous aider
              </h2>
              <ul style={{ width: "100%" }}>
                {results.slice(0, 3).map(({ _source: item }) => (
                  <li key={`${item.type}/${item.slug}`}>
                    <Link
                      href={{
                        pathname: `/${getRouteBySource(item.source)}/${
                          item.slug
                        }`,
                        query: { q: query, search: 0 },
                        hash: item.anchor
                      }}
                    >
                      <a>{item.title}</a>
                    </Link>
                  </li>
                ))}
              </ul>
            </React.Fragment>
          )}
        <h2 className="section__subtitle" style={{ alignSelf: "center" }}>
          Laissez-nous un commentaire sur notre service{" "}
          <span className="fontweight--normal fontweight--italic">
            {" "}
            - facultatif
          </span>
        </h2>
        <select
          name="motif"
          onChange={this.inputChange}
          onBlur={this.inputChange}
          className="feedback__input"
          value={this.state.motif}
        >
          {motifLabels.map((label, i) => (
            <option key={`motif-${i}`} value={label}>
              {label}
            </option>
          ))}
        </select>
        <textarea
          ref={this.texteareaRef}
          name="message"
          className="feedback__input"
          placeholder="Les informations ..."
          onChange={this.inputChange}
          value={this.state.message}
        />
        <input
          type="email"
          name="email"
          onChange={this.inputChange}
          placeholder="nom@adresse.email"
          className="feedback__input"
          value={this.state.email}
        />
        <div className="feedback__action">
          <button
            className="feedback__button btn btn__primary"
            disabled={this.state.status === "sending"}
          >
            Envoyer ma question
          </button>
          <div className="feedback__status"> {this.getAlert()} </div>
        </div>
      </form>
    );
  }
}

export { FeedbackForm };
