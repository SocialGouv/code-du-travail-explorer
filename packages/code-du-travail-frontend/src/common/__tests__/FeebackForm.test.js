import React from "react";
import { fireEvent, render, wait } from "react-testing-library";

import { FeedbackForm } from "../FeedbackForm";

jest.useFakeTimers();

const results = [
  {
    _source: {
      source: "code-du-travail",
      slug: "result-slug",
      title: "item title",
      type: "item-type"
    }
  }
];

describe("<FeedbackForm />", () => {
  it("should render", () => {
    const { container, getByValue } = render(
      <FeedbackForm
        query="Initial query"
        source="Tous contenus"
        results={results}
        onSubmit={jest.fn()}
      />
    );
    expect(getByValue("Initial query")).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it("should prevent submiting form if email is empty", () => {
    const { getByText } = render(
      <FeedbackForm
        query="Initial query"
        source="Tous contenus"
        results={results}
        onSubmit={jest.fn()}
      />
    );
    window.alert = jest.fn();
    getByText("Envoyer ma question").click();
    expect(alert).toHaveBeenCalled();
  });

  it("should submit form if email is filled", () => {
    const onSubmit = jest.fn().mockResolvedValue({ success: true });
    const { getByText, getByPlaceholderText } = render(
      <FeedbackForm
        query="Initial query"
        source="Tous contenus"
        results={results}
        onSubmit={onSubmit}
      />
    );
    const email = getByPlaceholderText("nom@adresse.email");
    const button = getByText("Envoyer ma question");

    fireEvent.change(email, { target: { value: "user@social.gouv.fr" } });
    button.click();
    expect(onSubmit).toHaveBeenCalledTimes(1);
  });

  it("should reset form after submit", async () => {
    const onSubmit = jest.fn().mockResolvedValue({ success: true });
    const { getByText, getByPlaceholderText } = render(
      <FeedbackForm
        query="Initial query"
        source="Tous contenus"
        results={results}
        onSubmit={onSubmit}
      />
    );

    const email = getByPlaceholderText("nom@adresse.email");
    const button = getByText(/Envoyer ma question/i);

    fireEvent.change(email, { target: { value: "user@social.gouv.fr" } });
    expect(email.value).toBe("user@social.gouv.fr");

    button.click();
    await wait(() => getByPlaceholderText("nom@adresse.email"));
    expect(email.value).toBe("");
  });

  it("should show a status message after submit", async () => {
    const onSubmit = jest.fn().mockResolvedValue({ success: true });
    const { getByText, getByPlaceholderText } = render(
      <FeedbackForm
        query="Initial query"
        source="Tous contenus"
        results={results}
        onSubmit={onSubmit}
      />
    );

    const email = getByPlaceholderText("nom@adresse.email");
    const button = getByText(/Envoyer ma question/i);

    fireEvent.change(email, { target: { value: "user@social.gouv.fr" } });
    button.click();

    await wait(() => expect(getByText(/message bien envoyé/i)).toBeTruthy());
  });

  it("should hide status message after 3s", async () => {
    const onSubmit = jest.fn().mockResolvedValue({ success: true });
    const { getByText, queryByText, getByPlaceholderText } = render(
      <FeedbackForm
        query="Initial query"
        source="Tous contenus"
        results={results}
        onSubmit={onSubmit}
      />
    );

    const email = getByPlaceholderText("nom@adresse.email");
    const button = getByText(/Envoyer ma question/i);

    fireEvent.change(email, { target: { value: "user@social.gouv.fr" } });
    button.click();
    await wait(() => queryByText(/message bien envoyé/i));
    jest.runAllTimers();
    expect(queryByText(/message bien envoyé/i)).not.toBeTruthy();
  });
  it("should show error status message if request fail ", async () => {
    const onSubmit = jest.fn().mockRejectedValue({});
    const { getByText, queryByText, getByPlaceholderText } = render(
      <FeedbackForm
        query="Initial query"
        source="Tous contenus"
        results={results}
        onSubmit={onSubmit}
      />
    );

    const email = getByPlaceholderText("nom@adresse.email");
    const button = getByText(/Envoyer ma question/i);
    fireEvent.change(email, { target: { value: "user@social.gouv.fr" } });
    button.click();
    await wait(() => expect(getByText(/Impossible/i)).toBeTruthy());
    jest.runAllTimers();
    expect(queryByText(/Impossible/i)).not.toBeTruthy();
  });
});
