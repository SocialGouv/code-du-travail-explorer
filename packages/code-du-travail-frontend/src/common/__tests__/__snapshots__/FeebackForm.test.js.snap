// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeedbackForm /> should render 1`] = `
<div>
  <form
    action="https://formspree.io/xbalrgzm"
    class="feedback__form"
  >
    <input
      class="feedback__input"
      disabled=""
      name="question"
      type="text"
      value="Initial query"
    />
    <input
      name="url"
      type="hidden"
      value="http://localhost/"
    />
    <h2
      class="section__subtitle"
    >
      Les réponses qui pourraient vous aider
    </h2>
    <ul
      style="width: 100%;"
    >
      <li>
        <a
          href="/undefined/result-slug?q=Initial%20query&search=0"
        >
          item title
        </a>
      </li>
    </ul>
    <h2
      class="section__subtitle"
    >
      Laissez-nous un commentaire sur notre service
       
      <span
        class="fontweight--normal fontweight--italic"
      >
         
        - facultatif
      </span>
    </h2>
    <select
      class="feedback__input"
      name="motif"
    >
      <option
        value="Les informations proposées ne sont pas à jour"
      >
        Les informations proposées ne sont pas à jour
      </option>
      <option
        value="Les informations me paraissent erronées"
      >
        Les informations me paraissent erronées
      </option>
      <option
        value="Les informations me paraissent incomplètes"
      >
        Les informations me paraissent incomplètes
      </option>
      <option
        value="Les informations me paraissent hors-sujet"
      >
        Les informations me paraissent hors-sujet
      </option>
      <option
        value="Je n'ai pas trouvé la réponse à ma question"
      >
        Je n'ai pas trouvé la réponse à ma question
      </option>
    </select>
    <textarea
      class="feedback__input"
      name="message"
      placeholder="Les informations ..."
    />
    <input
      class="feedback__input"
      name="email"
      placeholder="nom@adresse.email"
      type="email"
      value=""
    />
    <div
      class="feedback__action"
    >
      <button
        class="feedback__button btn btn__primary"
      >
        Envoyer ma question
      </button>
      <div
        class="feedback__status"
      >
         
         
      </div>
    </div>
  </form>
</div>
`;
